be/src/exec/vectorized/aggregate/agg_hash_variant.h

```diff
@@ -288,219 +449,46 @@ struct AggHashMapVariant {
-    std::unique_ptr<SerializedKeyFixedSize16AggHashMap<PhmapSeed2>> phase2_slice_fx16;
-
-    void init(RuntimeState* state, Type type_) {
-        type = type_;
-        switch (type_) {
-#define M(NAME)                                                                     \
-    case Type::NAME:                                                                \
-        NAME = std::make_unique<decltype(NAME)::element_type>(state->chunk_size()); \
-        break;
-            APPLY_FOR_AGG_VARIANT_ALL(M)
-#undef M
-        }
+    detail::AggHashMapWithKeyPtr hash_map_with_key;
+    auto& get_variant() { return hash_map_with_key; }
+    template <class Vistor>
+    auto visit(Vistor&& vistor) const {
+        return std::visit(std::forward<Vistor>(vistor), hash_map_with_key);
     }
 
+    void init(RuntimeState* state, Type type, AggStatistics* agg_statis);
+
+    void convert_to_two_level(RuntimeState* state);
+
     size_t capacity() const {
-        switch (type) {
-#define M(NAME)      \
-    case Type::NAME: \
-        return NAME->hash_map.capacity();
-            APPLY_FOR_AGG_VARIANT_ALL(M)
-#undef M
-        }
-        return 0;
+        return visit([](const auto& hash_map_with_key) { return hash_map_with_key->hash_map.capacity(); });
     }
 
     size_t size() const {
-        switch (type) {
-#define M(NAME)      \
-    case Type::NAME: \
-        return NAME->hash_map.size() + (NAME->null_key_data != nullptr);
-            APPLY_FOR_AGG_VARIANT_NULL(M)
-#undef M
-
-#define M(NAME)      \
-    case Type::NAME: \
-        return NAME->hash_map.size();
-            APPLY_FOR_AGG_VARIANT_NOT_NULL(M)
-#undef M
-        }
-        return 0;
+        return visit([](const auto& hash_map_with_key) {
+            return hash_map_with_key->hash_map.size() + (hash_map_with_key->get_null_key_data() != nullptr);
+        });
     }
 
     size_t reserved_memory_usage(const MemPool* pool) const {
-        switch (type) {
-#define M(NAME)      \
-    case Type::NAME: \
-        return NAME->hash_map.dump_bound() + pool->total_reserved_bytes();
-
-            APPLY_FOR_AGG_VARIANT_ALL(M)
-#undef M
-        }
-
-        return 0;
+        return visit([pool](const auto& hash_map_with_key) {
+            return hash_map_with_key->hash_map.dump_bound() + pool->total_reserved_bytes();
+        });
     }
 
     size_t allocated_memory_usage(const MemPool* pool) const {
-        switch (type) {
-#define M(NAME)      \
-    case Type::NAME: \
-        return sizeof(decltype(NAME)::element_type::KeyType) * NAME->hash_map.size() + pool->total_allocated_bytes();
-
-            APPLY_FOR_AGG_VARIANT_ALL(M)
-#undef M
-        }
-
-        return 0;
+        return visit([pool](const auto& hash_map_with_key) {
+            return sizeof(typename decltype(hash_map_with_key->hash_map)::key_type) *
+                           hash_map_with_key->hash_map.capacity() +
+                   pool->total_allocated_bytes();
```

Comment:
`allocated_memory_usage` is used to decide whether using passthrough or aggregation mode for streaming aggregation.
If we use `capacity` instead of `size`, will prefer passthrough mode than before?

Reply:
> allocated_memory_usage is used to decide whether using passthrough or aggregation mode for streaming aggregation.
If we use capacity instead of size, will prefer passthrough mode than before?

Yes, according my test it works for such query: (TPCDS-1T)
```
with 
 max_store_sales as
 (select max(csales) tpcds_cmax 
  from (select c_customer_sk,sum(cast(ss_quantity*ss_sales_price as float)) csales
        from store_sales
            ,customer
            ,date_dim 
        where ss_customer_sk = c_customer_sk
         and ss_sold_date_sk = d_date_sk
         and d_year in (2000,2000+1,2000+2,2000+3) 
        group by c_customer_sk) t1)
select count(*) from (
    select * from max_store_sales
)t;
```